ui/src/app/edge/history/common/energy/chart/chart.ts

```diff
@@ -101,7 +101,8 @@ export class ChartComponent extends AbstractHistoryChart {
             },
             converter: () =>
               data['ProductionActivePower']?.map((value, index) =>
-                value - data['GridSell'][index] - data['EssCharge'][index])?.map(value => HistoryUtils.ValueConverter.NEGATIVE_AS_ZERO(value)),
+                Utils.subtractSafely(Utils.subtractSafely(value, data['GridSell'][index]), data['EssCharge'][index]))
```

Comment:
I suggest changing the Utils method to accept a varargs parameter:

```
  public static subtractSafely(...values: number[]): number {
    var result = null;
    values.forEach(v => {
      if (v != null) {
        if (result == null) {
          result = v;
        } else {
          result -= v;
        }
      }
    })
    return result;
  }
```

then using it as 
```
            converter: () =>
              data['ProductionActivePower']?.map((value, index) => Utils.subtractSafely(value, data['GridSell'][index], data['EssCharge'][index]))
                ?.map(value => HistoryUtils.ValueConverter.NEGATIVE_AS_ZERO(value)),
```

Also there is a subtraction just above this line:
```
              return energyValues.result.data['_sum/ProductionActiveEnergy'] - energyValues.result.data['_sum/GridSellActiveEnergy'] - energyValues.result.data['_sum/EssDcChargeEnergy'];
```
Shouldn't you use Utils.subtractSafely there as well?

Reply:
Thanks for the insightful review, 
yes i should have used Utils.substractSafely there also.
i am working on the requested changes.